### ISSUE TITLE
Structured Bindings Miscompilation(?) on Windows x64.

### ISSUE THREAD

**Issue by 375f5d7a-e277-487e-b103-96457dd07676** [llvm/llvm-project#50075]
|  |  |
| --- | --- |
| Bugzilla Link | [50731](https://llvm.org/bz50731) |
| Version | 12.0 |
| OS | Windows NT |
| CC | @zygoloid |

## Extended Description 
Clang 11.0 and Clang 12.0 generates code that causes memory access violation (attempting to read address 0xffff...) on Windows x64 with `-O0` or `-O1` for the following code, https://godbolt.org/z/P6d3Pn1Gh

Note:
- Clang 10.0 doesn't cause this violation.
- Clang 12.0 with `-O2` or `-O3` doesn't cause this violation.
- Clang 12.0 doesn't cause this violation on Windows x86 regardless of optimization level.

Same as godbolt link:
```cpp
#include <immintrin.h>

#include <array>
#include <cassert>
#include <cstdint>
#include <vector>

template <class F>
void f(uint8_t const* a_ptr, uint8_t const* b_ptr, uint8_t const* c_ptr, uint8_t* d_ptr, size_t size, F f) noexcept {
  assert(size / 32 * 32 == size);

  for (auto x = static_cast<size_t>(0); x < size; x += 32) {
    auto a = _mm256_loadu_si256(reinterpret_cast<__m256i const*>(a_ptr + x));
    auto b = _mm256_loadu_si256(reinterpret_cast<__m256i const*>(b_ptr + x));
    auto c = _mm256_loadu_si256(reinterpret_cast<__m256i const*>(c_ptr + x));
    auto d = f({a, b, c});

    _mm256_storeu_si256(reinterpret_cast<__m256i*>(d_ptr + x), d);
  }
}

int main() {
  auto size = 32 * 32;

  auto a = std::vector<uint8_t>(size, 1);
  auto b = std::vector<uint8_t>(size, 2);
  auto c = std::vector<uint8_t>(size, 4);
  auto d = std::vector<uint8_t>(size);

  f(a.data(), b.data(), c.data(), d.data(), size, [](std::array<__m256i, 3> in) noexcept {
    auto [in0, in1, in2] = in;

    return _mm256_add_epi8(_mm256_add_epi8(in0, in1), in2);
  });

  return 0;
}
```

Changing,

```cpp
auto [in0, in1, in2] = in;
```

to,

```cpp
auto in0 = in[0];
auto in1 = in[1];
auto in2 = in[2];
```

seems to fix the problem for this example.

Also, Visual Studio (16.10.2) STL implementation of `std::array`, `std::get`, `std::tuple_size` seems to be implemented correctly.

**Comment by AaronBallman** [comment]
This issue appears to be fixed as of Clang 17, at least from my local testing. Closing the issue as resolved, but if the issue still reproduces for you, please reopen with more information.
